direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java

```diff
@@ -97,15 +98,21 @@
   private static final long IDLE_FLUSH_TIME = 500L;
   private static final long BOUNDED_OUT_OF_ORDERNESS = 5000L;
 
-  static class IntOffset implements Offset {
+  static class ConsumedOffset implements Offset {
 
     private static final long serialVersionUID = 1L;
 
-    @Getter final long offset;
+    static ConsumedOffset empty() {
+      return new ConsumedOffset(Collections.emptySet(), Long.MIN_VALUE);
+    }
+
     @Getter final long watermark;
+    @Getter final Set<String> consumedKeyAttr;
 
-    public IntOffset(long offset, long watermark) {
-      this.offset = offset;
+    public ConsumedOffset(final Set<String> consumedKeyAttr, long watermark) {
+      synchronized (consumedKeyAttr) {
+        this.consumedKeyAttr = new HashSet<>(consumedKeyAttr);
```

Comment:
java.util.Collections#unmodifiableSet

Reply:
Ah, actually no. That is not the same semantics. I need a clone.